tools/clang/unittests/HLSL/ExecutionTest.cpp

```diff
@@ -11330,6 +11343,124 @@ TEST_F(ExecutionTest, QuadAnyAll) {
     WEX::Logging::Log::Result(WEX::Logging::TestResults::Skipped);
 }
 
+TEST_F(ExecutionTest, IsNormalTest) {
+  EnableShaderBasedValidation();
+    WEX::TestExecution::SetVerifyOutput verifySettings(
+      WEX::TestExecution::VerifyOutputSettings::LogOnlyFailures);
+  CComPtr<IStream> pStream;
+  ReadHlslDataIntoNewStream(L"ShaderOpArith.xml", &pStream);
+
+  std::shared_ptr<st::ShaderOpSet> ShaderOpSet =
+      std::make_shared<st::ShaderOpSet>();
+  st::ParseShaderOpSetFromStream(pStream, ShaderOpSet.get());
+  st::ShaderOp *pShaderOp =
+      ShaderOpSet->GetShaderOp("IsNormal");
+  vector<st::ShaderOpRootValue> fallbackRootValues = pShaderOp->RootValues;
+
+  const int expectedResultsSize = 12;
+  // false represents 0, true represents a non-zero value
+  int expectedResults[expectedResultsSize] = {0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1};
+    
+  // TestShaderModels will be an array, where the first x models are "non-fallback", and the rest of the models
+  // are "fallback". If TestShaderModels has length y, and a test loops through all shader models, a convention
+  // to test based on whether fallback is enabled or not is to limit the loop like this:
+  // unsigned num_models_to_test = ExecutionTest::IsFallbackPathEnabled() ? y : x;
+  D3D_SHADER_MODEL sm = D3D_SHADER_MODEL_6_0;
+  LogCommentFmt(L"\r\nVerifying isNormal in shader "
+                L"model 6.%1u",
+                ((UINT)sm & 0x0f));
+
+  CComPtr<ID3D12Device> pDevice;
+  CreateDevice(&pDevice, sm, false /* skipUnsupported */);
+  D3D12_FEATURE_DATA_D3D12_OPTIONS devOptions;
+  VERIFY_SUCCEEDED(
+      pDevice->CheckFeatureSupport((D3D12_FEATURE)D3D12_FEATURE_D3D12_OPTIONS,
+                                    &devOptions, sizeof(devOptions)));
+  if (devOptions.ResourceBindingTier < D3D12_RESOURCE_BINDING_TIER_3) {
+    WEX::Logging::Log::Comment(
+        L"Device does not support Resource Binding Tier 3");
+    WEX::Logging::Log::Result(WEX::Logging::TestResults::Skipped);
+    return;
+  }
+
+  auto ShaderInitFn = [&](LPCSTR Name, LPCSTR pText, IDxcBlob **ppShaderBlob, st::ShaderOp *pShaderOp) {
```

Comment:
This function definition could be more generally useful if wrapped in a function that takes arguments for compilation and replacements, like so:
```
// Copies input strings to local storage, so it doesn't rely on lifetime of input string pointers.
st::ShaderOpTest::TShaderCallbackFn MakeShaderReplacementCallback(
    llvm::ArrayRef<LPCWSTR> dxcArgs,
    llvm::ArrayRef<LPCSTR> LookFors,
    llvm::ArrayRef<LPCSTR> Replacements,
    dxc::DxcDllSupport &dllSupport) {
  // place ArrayRef arguments in std::vector locals, and copy them by value into the callback function lambda
  std::vector<std::wstring> ArgsStorage(dxcArgs.size());
    for (unsigned i = 0; i < dxcArgs.size(); ++i)
      ArgsStorage[i] = dxcArgs[i];
  std::vector<std::string> LookForsStorage(LookFors.size());
    for (unsigned i = 0; i < LookFors.size(); ++i)
      LookForsStorage[i] = LookFors[i];
  std::vector<std::string> ReplacementsStorage(Replacements.size());
    for (unsigned i = 0; i < Replacements.size(); ++i)
      ReplacementsStorage[i] = Replacements[i];
  auto ShaderInitFn = 
      [ArgsStorage, LookForsStorage, ReplacementsStorage, &dllSupport]
      (LPCSTR Name, LPCSTR pText, IDxcBlob **ppShaderBlob, st::ShaderOp *pShaderOp) {
    // Create pointer vectors from local storage to supply API needs
    std::vector<LPCWSTR> Args(ArgsStorage.size());
    for (unsigned i = 0; i < ArgsStorage.size(); ++i)
      args[i] = ArgsStorage[i].c_str();
    std::vector<LPCSTR> LookFors(LookForsStorage.size());
    for (unsigned i = 0; i < LookForsStorage.size(); ++i)
      LookFors[i] = LookForsStorage[i].c_str();
    std::vector<LPCSTR> Replacements(ReplacementsStorage.size());
    for (unsigned i = 0; i < ReplacementsStorage.size(); ++i)
      Replacements[i] = ReplacementsStorage[i].c_str();
    // Do the rest of the work for compilation, replacement, and root sig copying, if present
  }
  return ShaderInitFn;
}
```